source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,209 @@
+        std::regex_search(headers.AcceptEncoding()->value().c_str(), acceptEncodingRegex()));
+  return FilterHeadersStatus::Continue;
+}
+
+FilterHeadersStatus GzipFilter::encodeHeaders(HeaderMap& headers, bool end_stream) {
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isContentTypeAllowed(headers) && isCacheControlAllowed(headers) && isEtagAllowed(headers) &&
+      isLastModifiedAllowed(headers) && isTransferEncodingAllowed(headers) &&
+      !headers.ContentEncoding()) {
+    // TODO(gsagula): tests needed.
+    config_->disableVary()
+        ? headers.removeVary()
+        : headers.insertVary().value(Http::Headers::get().VaryValues.AcceptEncoding);
+
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(Http::Headers::get().ContentEncodingValues.Gzip);
+    compressor_.init(config_->compressionLevel(), config_->compressionStrategy(),
+                     config_->windowBits(), config_->memoryLevel());
+  } else {
+    skip_compression_ = true;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_compression_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t n_data{data.length()};
+
+  if (n_data) {
+    compressor_.compress(data, compressed_data_);
+  }
+
+  if (end_stream) {
+    compressor_.flush(compressed_data_);
+  }
+
+  if (compressed_data_.length()) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::isMinimumContentLength(HeaderMap& headers) const {
+  const Http::HeaderEntry* content_length = headers.ContentLength();
+  if (content_length) {
+    uint64_t length{};
+    StringUtil::atoul(content_length->value().c_str(), length);
+    return length >= config_->minimumLength();
+  }
+
+  const Http::HeaderEntry* transfer_encoding = headers.TransferEncoding();
+  return (transfer_encoding && transfer_encoding->value().find(
+                                   Http::Headers::get().TransferEncodingValues.Chunked.c_str()));
+}
+
+bool GzipFilter::isContentTypeAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type && !config_->contentTypeValues().empty()) {
+    std::string content_type_str{content_type->value().c_str()};
+    StringUtil::eraseFrom(content_type_str, ";");
+    StringUtil::trim(content_type_str);
+    return config_->contentTypeValues().find(content_type_str) !=
+           config_->contentTypeValues().end();
+  }
+
+  return true;
+}
+
+bool GzipFilter::isCacheControlAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  return cache_control ? !cache_control->value().find(
+                             Http::Headers::get().CacheControlValues.NoTransform.c_str())
```

Comment:
again this find() methodology will match substrings.  I'm just being paranoid.  Maybe none of these would actually ever yield a false match, but you do this pattern so much I think it might be better to add a helper like:
 
```
   StringUtil::containsToken(absl::string_view delim_separated_tokens, absl::string_view delimiters, bool trim_whitespace, absl::string_view token);
```

wdyt?

Reply:
I think that's the right approach! As per my comment above, I will review all instances that I used find. Thanks for pushing on this change.